File: Python/020_ValidParentheses.py
## 020 Valid Parentheses
## Given a string containing just the characters '(', ')', '{', '}', '[' and ']', 
## determine if the input</br> string is valid.
## The brackets must close in the correct order, "()" and "()[]{}" are all valid 
## but "(]" and "([)]" are not.

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        test = list()
        for key, val in enumerate(s):
            if val in ['(', '{', '[']:
                test.append(val)
            elif len(test) == 0:
                return False
            else:
                c = test.pop()
                if {')':'(', '}':'{', ']':'['}[val] != c: #这是一个字典索引
                    return False
        if len(test) == 0:
            return True
        else:
            return False

File: Cpp/020_ValidParentheses.cpp
/*
## 020 Valid Parentheses
## Given a string containing just the characters '(', ')', '{', '}', '[' and ']', 
## determine if the input</br> string is valid.
## The brackets must close in the correct order, "()" and "()[]{}" are all valid 
## but "(]" and "([)]" are not.
*/

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        char c;
        for(int i = 0; i < s.size(); i++)
        {
            if(s[i] == '(' || s[i] == '{' ||
               s[i] == '[')
                st.push(s[i]);
            else
            {
                if(st.empty()) //前面的已经匹配完毕，则若栈为空，必定无法匹配
                    return false;
                else
                {
                    c = st.top();
                    st.pop();
                    if((c != '(' && s[i] == ')') ||
                       (c != '[' && s[i] == ']') ||
                       (c != '{' && s[i] == '}'))
                        return false;                        
                }
            }
                
        }
        return st.empty();
        
    }
};